### PR TITLE
Updates to support nile main, add unit tests

### LINKED ISSUE #15: Update entry points format
Depends on https://github.com/OpenZeppelin/nile/pull/273

### LINKED ISSUE #18: Add unit tests
I propose adding unit tests to ensure proper internal function calls, error handling, raised errors, logging, etc.

### LINKED ISSUE #20: Remove max_fee argument from test script
Remove max_fee from https://github.com/OpenZeppelin/openzeppelin-nile-upgrades/blob/main/tests/resources/scripts/deploy_upgrade_proxy.py (pending next release of Nile with https://github.com/OpenZeppelin/nile/pull/262)

### LINKED ISSUE #21: Expect `get_hash` to return hex
Expect `get_hash` to return hex in `declare-impl`

### PR DESCRIPTION
**Pull request by ericglau**
Fixes #15 
Fixes #18 
Fixes #20 
Fixes #21 

Notable changes:
- Changed API to support NRE only, removed CLI
- Improved hash and address handling according to [Nile's hex/int pattern](https://github.com/OpenZeppelin/nile/blob/main/CONTRIBUTING.md#hexint-pattern)
- Added unit tests

### CODE REVIEW

**Comment by ericglau** [comment]
@andrew-fleming @ericnordelo @martriay Can you please review?

**Review by andrew-fleming — COMMENTED** [review]
Looking really good, @ericglau! I left a few suggestions and comments :)

File: src/nile_upgrades/upgrade_proxy.py [hunk 1]
```diff
@@ -1,53 +1,35 @@
...
-@click.command()
-@click.argument("signer", type=str)
-@click.argument("proxy_address_or_alias", type=str)
-@click.argument("contract_name", type=str)
-@click.option(
-    "--max_fee", nargs=1, help="Maximum fee for the transaction. Defaults to 0."
-)
-def upgrade_proxy(signer, proxy_address_or_alias, contract_name, max_fee=None):
+def upgrade_proxy(
+    nre, signer, proxy_address_or_alias, contract_name, max_fee=None, standalone_mode=None
```
[the remark this answers is not in the document]
Comment (ericglau) [reply src/nile_upgrades/upgrade_proxy.py:12]: Opened https://github.com/OpenZeppelin/nile/issues/295

File: tests/test_common.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+
+@patch("nile_upgrades.common.get_hash", return_value=PADDED_HASH)
+@patch("nile_upgrades.common.class_hash_exists", return_value=True)
+def test_declare_impl_already_exists(
+    mock_class_hash_exists, mock_get_hash, caplog
+):
+    logging.getLogger().setLevel(logging.DEBUG)
+
+    result = declare_impl(NETWORK, CONTRACT, SIGNER, None);
```
[the remark this answers is not in the document]
Comment (ericglau) [reply tests/test_common.py:26]: Added `conftest.py`

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -17,8 +17,13 @@ jobs:
         python_version: [ '3.8', '3.9' ]
         os: [ubuntu-latest]
     name: ${{ matrix.os }} - Python ${{ matrix.python_version }}
-    steps:
-      - uses: actions/checkout@v1
+    steps:   
+      - uses: actions/checkout@v3
+
+      - uses: actions/checkout@v3
```
[the remark this answers is not in the document]
Comment (ericglau) [reply .github/workflows/ci.yml:23]: It's needed twice to checkout different repos (this repo and Nile separately)

[the remark this answers is not in the document]
Comment (andrew-fleming) [reply .github/workflows/ci.yml:23]: Oh right!

File: tests/test_common.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+@pytest.mark.asyncio
+@patch("nile_upgrades.common.get_hash", return_value=PADDED_HASH)
+@patch("nile_upgrades.common.class_hash_exists", return_value=True)
+async def test_declare_impl_already_exists(
+    mock_class_hash_exists, mock_get_hash, caplog
+):
+    logging.getLogger().setLevel(logging.DEBUG)
+
+    result = await declare_impl(NETWORK, CONTRACT, SIGNER, None);
```
Comment (andrew-fleming) [inline tests/test_common.py:27]:
```suggestion
    result = await declare_impl(NETWORK, CONTRACT, SIGNER, None)
```

File: tests/test_common.py [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+
+@pytest.mark.asyncio
+@patch("nile_upgrades.common.get_hash", return_value=PADDED_HASH)
+@patch("nile_upgrades.common.class_hash_exists", return_value=False)
+@patch("nile.core.account.Account.declare", return_value=WRONG_HASH)
+async def test_declare_impl_not_match(
+    mock_declare, mock_class_hash_exists, mock_get_hash, caplog
+):
+    with pytest.raises(Exception) as e:
+        await declare_impl(NETWORK, CONTRACT, SIGNER, None);
```
Comment (andrew-fleming) [inline tests/test_common.py:42]:
```suggestion
        await declare_impl(NETWORK, CONTRACT, SIGNER, None)
```

File: tests/test_common.py [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+@pytest.mark.asyncio
+@patch("nile_upgrades.common.get_hash", return_value=PADDED_HASH)
+@patch("nile_upgrades.common.class_hash_exists", return_value=False)
+@patch("nile.core.account.Account.declare", return_value=PADDED_HASH)
+async def test_declare_impl(
+    mock_declare, mock_class_hash_exists, mock_get_hash, caplog
+):
+    logging.getLogger().setLevel(logging.DEBUG)
+
+    result = await declare_impl(NETWORK, CONTRACT, SIGNER, None);
```
Comment (andrew-fleming) [inline tests/test_common.py:55]:
```suggestion
    result = await declare_impl(NETWORK, CONTRACT, SIGNER, None)
```

File: tests/test_deploy_proxy.py [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+@patch("nile_upgrades.deploy_proxy.declare_impl", return_value=CLASS_HASH)
+@patch("nile_upgrades.deploy_proxy.get_selector_from_name", return_value=SELECTOR)
+@patch("nile_upgrades.deploy_proxy._get_proxy_artifact_path", return_value=PROXY_ARTIFACT_PATH)
+async def test_deploy_proxy(
+    mock_get_proxy_artifact_path, mock_get_selector, mock_declare_impl, caplog
+):
+    logging.getLogger().setLevel(logging.DEBUG)
+
+    with patch("nile.nre.NileRuntimeEnvironment.deploy", new=AsyncMock(return_value=(PROXY_ADDR_INT, IMPL_ABI))) as mock_deploy:
+        result = await deploy_proxy(NileRuntimeEnvironment(), SIGNER, CONTRACT, ARGS);
```
Comment (andrew-fleming) [inline tests/test_deploy_proxy.py:39]:
```suggestion
        result = await deploy_proxy(NileRuntimeEnvironment(), SIGNER, CONTRACT, ARGS)
```

File: tests/test_deploy_proxy.py [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+@patch("nile_upgrades.deploy_proxy.declare_impl", return_value=CLASS_HASH)
+@patch("nile_upgrades.deploy_proxy.get_selector_from_name", return_value=SELECTOR)
+@patch("nile_upgrades.deploy_proxy._get_proxy_artifact_path", return_value=PROXY_ARTIFACT_PATH)
+async def test_deploy_proxy_all_opts(
+    mock_get_proxy_artifact_path, mock_get_selector, mock_declare_impl, caplog
+):
+    logging.getLogger().setLevel(logging.DEBUG)
+
+    with patch("nile.nre.NileRuntimeEnvironment.deploy", new=AsyncMock(return_value=(PROXY_ADDR_INT, IMPL_ABI))) as mock_deploy:
+        result = await deploy_proxy(NileRuntimeEnvironment(), SIGNER, CONTRACT, ARGS, CUSTOM_INIT, ALIAS, MAX_FEE);
```
Comment (andrew-fleming) [inline tests/test_deploy_proxy.py:55]:
```suggestion
        result = await deploy_proxy(NileRuntimeEnvironment(), SIGNER, CONTRACT, ARGS, CUSTOM_INIT, ALIAS, MAX_FEE)
```

**Review by andrew-fleming — COMMENTED** [review]
Looking good! I left a few super minor suggestions to remove semicolons. 

Also, when I run `pytest`, I'm getting this error:

`'HTTP error ocurred. Status: 500. Text: <!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 3.2 Final//EN">\n<title>500 Internal S...nd was unable to complete your request. Either the server is overloaded or there is an error in the application.</p>\n'`

Tests run fine when the node is running, however. I'm thinking it might be a good idea to mock the account deployments in the tests after all. We did this in Nile once we started receiving `BadRequest` from the tests calling goerli with the exception of testing the node. I suppose that same exception works here with the integration test.

If the tests work when they're ran from the CI, maybe it's worth pushing it later. Or maybe it's my machine? Thoughts?

**Comment by ericglau** [comment]
I'm not sure what is the cause of the 500 errors.  The integration test starts the node and it passes consistently on my machine and on the CI, so I think that can be addressed later.

**Comment by ericglau** [comment]
The 500 error occurred for me if there's already a stale instance of Nile running on port 5000 in the background.  The workaround is to kill processes on port 5000 first.

**Pull request merged**